File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 0;
        Map<Character, Integer> map = new HashMap<>(); // to store frequency of characters
        int i = 0; // sliding window left pointer
        for (int j = 0; j < s.length(); j++) { // right pointer of sliding window
            char endOfWindow = s.charAt(j);
            map.put(endOfWindow, map.getOrDefault(endOfWindow, 0) + 1);
            while (map.get(endOfWindow) > 1) { // when not valid window
                char startOfWindow = s.charAt(i);
                map.put(startOfWindow, map.get(startOfWindow) - 1); // make it valid by incrementing the left pointer and decreasing the char's(which is repeating) count as well
                i++;
            }
            res = Math.max(res, j - i + 1); // update length of the valid substring(in the question's context)
            // Taking max as we need length of the largest valid substring
        }
        return res;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        
        unordered_map<char, int> map;
        int res = 0;
        
        for (int i = 0, j = 0; j < n; j++) {
            map[s[j]]++;
            while (map[s[j]] > 1) { // move the window to the right, as we move decrement the count of the chars in the window
                --map[s[i++]];
            }
            res = max(res, j - i + 1);
        }
        return res;
    }
};